aggregation/simple/src/main/java/com/spotify/heroic/aggregation/simple/RateInstance.java

```diff
@@ -0,0 +1,176 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.heroic.aggregation.simple;
+
+import com.spotify.heroic.ObjectHasher;
+import com.spotify.heroic.aggregation.AggregationInstance;
+import com.spotify.heroic.aggregation.AggregationOutput;
+import com.spotify.heroic.aggregation.AggregationResult;
+import com.spotify.heroic.aggregation.AggregationSession;
+import com.spotify.heroic.aggregation.BucketStrategy;
+import com.spotify.heroic.aggregation.EmptyInstance;
+import com.spotify.heroic.aggregation.RetainQuotaWatcher;
+import com.spotify.heroic.common.DateRange;
+import com.spotify.heroic.common.Series;
+import com.spotify.heroic.metric.Event;
+import com.spotify.heroic.metric.MetricCollection;
+import com.spotify.heroic.metric.MetricGroup;
+import com.spotify.heroic.metric.MetricType;
+import com.spotify.heroic.metric.Payload;
+import com.spotify.heroic.metric.Point;
+import com.spotify.heroic.metric.Spread;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import lombok.Data;
+
+/**
+ * A rate implementation inspired by the prometheus rate() function.
+ *
+ * It is intended to be be used with counters that are increasing to calculate the difference
+ * between two points. It handles the counter being reset.
+ *
+ * @author dmichel
+ */
+@Data
+public class RateInstance implements AggregationInstance {
+    private static final EmptyInstance INNER = EmptyInstance.INSTANCE;
+
+    @Override
+    public long estimate(DateRange range) {
+        return INNER.estimate(range);
+    }
+
+    @Override
+    public long cadence() {
+        return -1;
+    }
+
+    @Override
+    public AggregationInstance distributed() {
+        return this;
+    }
+
+    @Override
+    public AggregationInstance reducer() {
+        return INNER;
+    }
+
+    @Override
+    public boolean distributable() {
+        return false;
+    }
+
+    @Override
+    public void hashTo(final ObjectHasher hasher) {
+        hasher.putObject(getClass());
+    }
+
+    List<Point> computeDiff(List<Point> points) {
+        final Iterator<Point> it = points.iterator();
+        final ArrayList<Point> result = new ArrayList<>();
+
+        if (!it.hasNext()) {
+            return Collections.emptyList();
+        }
+
+        Point previous = it.next();
+
+        while (it.hasNext()) {
+            final Point current = it.next();
+
+            // Counter reset.
+            if (current.getValue() < previous.getValue()) {
```

Comment:
This doesn't necessarily mean the counter was reset. Counters can be decremented, and I think this won't calculate the rate correctly in the case.

Reply:
you are right i'll fix this comment...

in the docs for this aggregation I will explain this should only be used with monotonically increasing counters. this is the same as prom https://github.com/prometheus/prometheus/blob/2bd510a63e48ac6bf4971d62199bdb1045c93f1a/docs/querying/functions.md#rate